as-ps/as_ps/main.py

```diff
@@ -77,17 +75,29 @@ def bbblist(self):
     def process(self):
         """Process all read and write requests in queue."""
         t0 = _time.time()
-        if self._prucqueue:
-            self._prucqueue.process()
-        for bbb in self.bbblist:
-            self._scan_bbb(bbb)
-        self.driver.updatePVs()
-        t1 = _time.time()
-        # TODO: this detailed time keeping was necessary in order to
-        # have an update refresh rate at 10 Hz. scan_bbb is taking around
-        # 40 ms to complete. We should start optimizing
-        # the IOC code. As it is it is taking up 80% of BBB1 cpu time.
-        _time.sleep(abs(self._interval-(t1-t0)))
+        if self._prucqueue:  # Not None and not empty
+            status = self._prucqueue.process()
+            if status:
+                txt = ("[{:.2s}] - new thread started for write queue item. "
+                       "items left: {}")
+                logmsg = txt.format('Q ', len(self._prucqueue))
+                _log.info(logmsg)
+            # scanning bbb allows for read-only variables to be updated while
+            # there are pending write operations in the queue, since setpoint
+            # variables will not be updated in the scan, in order not to spoil
+            # the received write value.
+            for bbb in self.bbblist:
+                self._scan_bbb(bbb)
+            self.driver.updatePVs()
+            _time.sleep(0.050)
+        else:
+            for bbb in self.bbblist:
+                self._scan_bbb(bbb)
+            self.driver.updatePVs()
```

Comment:
I think it is better to call `self.driver.updatePV` inside `self._update_ioc_database` instead of calling `self.driver.updatePVs` here, for performance and code style reasons.

Reply:
For performance maybe it is better to call it in `self._scan_bbb`, since it does update PVs one time for all devices.